### ISSUE TITLE
Memory leak using fit_generator

### ISSUE THREAD

**Issue by nicolamarinello** [keras-team/keras#12100]
I'm developing a project to train a CNN classifier with images that are distributed across multiple .h5 files. The training dataset is quite big then I'm using a generator in order to read data from disk and feed the network. However as soon as the training starts, the used memory (RAM) starts increasing very fast until it gets completely filled and the training process gets stuck without any error.

At the beginning I thought the problem was related to h5py or the callbacks but then I removed all the unnecessary things to make things simpler and this phenomena still happens.

My configuration is Ubuntu 18.04, Keras from branch master 8c9de83, TensorFlow 1.12, CUDA 10.0 and cuDNN 7.

However this seems not to happen on CentOS with CPU training. 

In the following I will post the code to reproduce the issue. It is a simplified version of the original one but it easily reproduce the problem (the network should achieve 0.5 of accuracy as all the data is randomly generated)

```python
training_generator = DataGeneratorC(dataset_len=100000, batch_size=batch_size, shuffle=shuffle)

model = Sequential()

# simple network
model.add(Conv2D(32, kernel_size=(3, 3), activation='relu', input_shape=(1, img_rows, img_cols), data_format='channels_first'))
model.add(Conv2D(64, (3, 3), data_format='channels_first', activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2), data_format='channels_first'))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(4, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(units=1, activation='sigmoid'))

model.summary()

sgd = optimizers.SGD(lr=0.1, decay=1e-4, momentum=0.9, nesterov=True)

model.compile(optimizer=sgd, loss='binary_crossentropy', metrics=['accuracy'])
    
model.fit_generator(generator=training_generator, epochs=epochs, verbose=1, use_multiprocessing=True, workers=workers, shuffle=False)

```

Just a generator that feeds the network with random data

```python
import keras
import numpy as np
import h5py
import multiprocessing


class DataGeneratorC(keras.utils.Sequence):
    'Generates data for Keras'
    def __init__(self, dataset_len, batch_size=32, shuffle=True):
        self.batch_size = batch_size
        # self.h5files = h5files
        self.indexes = np.array([], dtype=np.int64).reshape(0, 1)
        self.shuffle = shuffle
        self.n_images = dataset_len
        self.generate_indexes()
        self.on_epoch_end()

    def __len__(self):
        'Denotes the number of batches per epoch'
        return int(np.floor(self.n_images/self.batch_size))

    def __getitem__(self, index):

        indexes = self.indexes[index * self.batch_size:(index + 1) * self.batch_size]

        # Generate data
        x, y = self.__data_generation(indexes)

        return x, y

    def get_indexes(self):
        return self.indexes

    def generate_indexes(self):
         # indexing of the dataset in some way
         # self.indexes in some way
         self.indexes = np.arange(self.n_images)

    def on_epoch_end(self):
        'Updates indexes after each epoch'
        if self.shuffle:
            np.random.shuffle(self.indexes)

    def __data_generation(self, indexes):
        'Generates data containing batch_size samples'
        # Initialization
        x = np.empty([self.batch_size, 100, 100])
        y = np.empty([self.batch_size], dtype=int)

        # Generate data
        for i, row in enumerate(indexes):
            # project code
            # filename = self.h5files[row[0]]
            # h5f = h5py.File(filename, 'r')
            # Store image
            # x[i, ] = h5f['path/path_images'][row[1]]
            # h5f.close()
            # Store class
            # y[i] = row[2]

            # test code
            y[i] = np.random.randint(2, size=1) # this returns 0 or 1
            x[i, ] = np.random.rand(100, 100)

        x = x.reshape(x.shape[0], 1, 100, 100)

        return x, y
```

**Comment by gabrieldemarmiesse** [comment]
Thanks for giving a small reproducible example, this should help with debugging. @Dref360 Do you have any idea what that may be? At first glance I don't see any problem with the python code so this memory error is definitly strange.

**Comment by Dref360** [comment]
Cannot reproduce on my setup. Using keras master. Displaying the % of memory at each step. It stabilizes around 6% after some epochs.

```python
import os

import numpy as np
import psutil

import keras
from keras import Sequential, optimizers
from keras.callbacks import Callback
from keras.layers import Conv2D, MaxPooling2D, Dropout, Flatten, Dense


class DataGeneratorC(keras.utils.Sequence):
    'Generates data for Keras'

    def __init__(self, dataset_len, batch_size=32, shuffle=True):
        self.batch_size = batch_size
        # self.h5files = h5files
        self.indexes = np.array([], dtype=np.int64).reshape(0, 1)
        self.shuffle = shuffle
        self.n_images = dataset_len
        self.generate_indexes()
        self.on_epoch_end()

    def __len__(self):
        'Denotes the number of batches per epoch'
        return int(np.floor(self.n_images / self.batch_size))

    def __getitem__(self, index):

        indexes = self.indexes[index * self.batch_size:(index + 1) * self.batch_size]

        # Generate data
        x, y = self.__data_generation(indexes)

        return x, y

    def get_indexes(self):
        return self.indexes

    def generate_indexes(self):
        # indexing of the dataset in some way
        # self.indexes in some way
        self.indexes = np.arange(self.n_images)

    def on_epoch_end(self):
        'Updates indexes after each epoch'
        if self.shuffle:
            np.random.shuffle(self.indexes)

    def __data_generation(self, indexes):
        'Generates data containing batch_size samples'
        # Initialization
        x = np.empty([self.batch_size, 100, 100])
        y = np.empty([self.batch_size], dtype=int)

        # Generate data
        for i, row in enumerate(indexes):
            x[i,] = np.random.rand(100, 100)

        x = x.reshape(x.shape[0], 1, 100, 100)

        return x, y


batch_size = 30
shuffle = True

training_generator = DataGeneratorC(dataset_len=100000, batch_size=batch_size, shuffle=shuffle)

model = Sequential()

# simple network
model.add(Conv2D(32, kernel_size=(3, 3), activation='relu', input_shape=(1, 100, 100),
                 data_format='channels_first'))
model.add(Conv2D(64, (3, 3), data_format='channels_first', activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2), data_format='channels_first'))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(4, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(units=1, activation='sigmoid'))

model.summary()

sgd = optimizers.SGD(lr=0.1, decay=1e-4, momentum=0.9, nesterov=True)

model.compile(optimizer=sgd, loss='binary_crossentropy', metrics=['accuracy'])


class MyCallback(Callback):
    def on_batch_end(self, batch, logs):
        process = psutil.Process(os.getpid())
        print(process.memory_percent())


model.fit_generator(generator=training_generator, epochs=100, verbose=0, use_multiprocessing=True, workers=4,
                    shuffle=False, callbacks=[MyCallback()])

```

**Comment by gabrieldemarmiesse** [comment]
Thanks a lot @Dref360 for taking the time to try to reproduce this! I'll be closing this issue then. 

@nicolamarinello This is very likely an issue with your setup. Even if it isn't, we can't flush out a bug that we can't see in our own machines. This is strange that the issue goes away when you switch of OS, I believe it's another sign that it's unrelated to the keras codebase.

**Comment by nicolamarinello** [comment]
Have you tried to check with ```free -m``` during the training? If I check the output of 

```python
print(process.memory_percent())
```
I see that the used memory is almost constant, but checking with ```free -m``` or ```htop``` I see that the used memory keep increasing until the end of the epoch, then the memory is cleaned and it starts again during the next epoch.

**Comment by Dref360** [comment]
Still stable on my side.

**Comment by nicolamarinello** [comment]
@Dref360 which setup are you using?

**Comment by Dref360** [comment]
Ubuntu 18.04, Anaconda Python 3.6, Keras master, TF 1.12.0, Numpy 1.15.4

**Comment by Dref360** [comment]
Right now CPU, I will test on GPU tonight

**Comment by sinclairnick** [comment]
Just leaving this here for future people like myself. To fix a memory leak I removed the TensorBoard callback. #34876 also references this

**Comment by wingdi** [comment]
@nicolamarinello  Do you solve this problem finally ?  I have the same problem like this now .

**Comment by keivanB** [comment]
same here

**Comment by junliu-cn** [comment]
same here

**Comment by cmantas** [comment]
there seem to be a few similar issues
https://github.com/tensorflow/tensorflow/issues/31312
https://github.com/tensorflow/tensorflow/issues/50379
